src/client/renderer/r_types.h

```diff
@@ -94,6 +94,99 @@ typedef enum {
 #define R_BUFFER_TYPE_MASK 0x03
 
 /**
+ * @brief Attribute indices - these should be assigned to
+ * every program that uses them, and are also used for buffer storage.
+ */
+typedef enum {
+	R_ARRAY_POSITION,
+	R_ARRAY_COLOR,
+	R_ARRAY_NORMAL,
+	R_ARRAY_TANGENT,
+	R_ARRAY_DIFFUSE,
+	R_ARRAY_LIGHTMAP,
+
+	// These three are only used for shader-based lerp.
+	// They are only enabled if the ones that match up
+	// to it are enabled as well.
+	R_ARRAY_NEXT_POSITION,
+	R_ARRAY_NEXT_NORMAL,
+	R_ARRAY_NEXT_TANGENT,
+
+	R_ARRAY_MAX_ATTRIBS,
+	R_ARRAY_ALL = R_ARRAY_MAX_ATTRIBS,
+
+	// This is a special entry so that R_BindAttributeBuffer can be
+	// used for binding element buffers as well.
+	R_ARRAY_ELEMENTS = -1
+} r_attribute_id_t;
+
+/**
+ * @brief These are the masks used to tell which data
+ * should be actually bound. They should match
+ * up with the ones above to make things simple.
+ */
+typedef enum {
+	R_ARRAY_MASK_POSITION		= (1 << R_ARRAY_POSITION),
+	R_ARRAY_MASK_COLOR			= (1 << R_ARRAY_COLOR),
+	R_ARRAY_MASK_NORMAL			= (1 << R_ARRAY_NORMAL),
+	R_ARRAY_MASK_TANGENT		= (1 << R_ARRAY_TANGENT),
+	R_ARRAY_MASK_DIFFUSE		= (1 << R_ARRAY_DIFFUSE),
+	R_ARRAY_MASK_LIGHTMAP		= (1 << R_ARRAY_LIGHTMAP),
+
+	R_ARRAY_MASK_NEXT_POSITION	= (1 << R_ARRAY_NEXT_POSITION),
+	R_ARRAY_MASK_NEXT_NORMAL	= (1 << R_ARRAY_NEXT_NORMAL),
+	R_ARRAY_MASK_NEXT_TANGENT	= (1 << R_ARRAY_NEXT_TANGENT),
+
+	R_ARRAY_MASK_ALL			= (1 << R_ARRAY_MAX_ATTRIBS) - 1
+} r_attribute_mask_t;
+
+typedef enum {
+	R_ATTRIB_FLOAT,
+	R_ATTRIB_BYTE,
+	R_ATTRIB_UNSIGNED_BYTE,
+	R_ATTRIB_SHORT,
+	R_ATTRIB_UNSIGNED_SHORT,
+	R_ATTRIB_INT,
+	R_ATTRIB_UNSIGNED_INT,
+	R_ATTRIB_INT_2_10_10_10_REV,
+
+	R_ATTRIB_TOTAL_TYPES
+} r_attrib_type_t;
+
+// NOTE: if any of the enums or data types
```

Comment:
I'd rather see Doxygen `@brief` `@details` and `@remarks` comments for stuff like this, and 100 char line wrap. Nothing to hold up the pull request for. Just a standard I'd like to move to for block-level comments.